Documentation/Proposals/NNNN-custom-test-execution-traits.md

```diff
@@ -0,0 +1,397 @@
+  /// arbitrary work before or after invoking `function`.
+  ///
+  /// This method should either invoke `function` once before returning or throw
+  /// an error if it is unable to perform its custom logic successfully.
+  ///
+  /// This method is invoked once for the test its associated trait is applied
+  /// to, and then once for each test case in that test, if applicable. If a
+  /// test is skipped, this method is not invoked for that test or its cases.
+  ///
+  /// Issues recorded by this method are associated with `test`.
+  func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws
+}
+
+public protocol Trait: Sendable {
+  // ...
+
+  /// The type of the test executor for this trait.
+  ///
+  /// The default type is `Never`, which cannot be instantiated. This means the
+  /// value of the ``testExecutor-714gp`` property for all traits with
+  /// the default custom executor type is `nil`, meaning such traits will not
+  /// perform any custom execution for the tests they're applied to.
+  associatedtype TestExecutor: TestExecuting = Never
+
+  /// The test executor for this trait, if any.
+  ///
+  /// If this trait's type conforms to ``TestExecuting``, the default value of
+  /// this property is `self` and this trait will be used to customize test
+  /// execution. This is the most straightforward way to implement a trait which
+  /// customizes the execution of tests.
+  ///
+  /// If the value of this property is an instance of a different type
+  /// conforming to ``TestExecuting``, that instance will be used to perform
+  /// test execution instead.
+  ///
+  /// The default value of this property is `nil` (with the default type
+  /// `Never?`), meaning that instances of this type will not perform any custom
+  /// test execution for tests they are applied to.
+  var testExecutor: TestExecutor? { get }
+}
+
+extension Trait where Self: TestExecuting {
+  // Returns `self`.
+  public var testExecutor: Self? {
+}
+
+extension Trait where TestExecutor == Never {
+  // Returns `nil`.
+  public var testExecutor: TestExecutor? {
+}
+
+extension Never: TestExecuting {}
+```
+
+Here is a complete example of the usage scenario described earlier, showcasing
+the proposed APIs:
+
+```swift
+@Test(.mockAPICredentials)
+func example() {
+  // ...validate API usage, referencing `APICredentials.current`...
+}
+
+struct MockAPICredentialsTrait: TestTrait, TestExecuting {
+  func execute(_ function: @Sendable () async throws -> Void, for test: Test, testCase: Test.Case?) async throws {
+    let mockCredentials = APICredentials(apiKey: "...")
+    try await APICredentials.$current.withValue(mockCredentials) {
+      try await function()
+    }
+  }
+}
+
+extension Trait where Self == MockAPICredentialsTrait {
+  static var mockAPICredentials: Self {
+    Self()
+  }
+}
+```
+
+## Source compatibility
+
+The proposed APIs are purely additive.
```

Comment:
Might be worth a note here that this change will obsolete the `SPIAwareTrait` and `CustomExecutionTrait` experimental SPI.

Reply:
Noted!